### PR TITLE
[DeviceMesh] Move DeviceMesh out from torch.distributed._tensor

### PR DESCRIPTION
**Pull request by wz337**
Move DeviceMesh out as a standalone module. Once we make sure everything is migrated and doc is ready, we will make `torch.distributed._device_mesh` public in follow-up PRs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Review by wanchaol — CHANGES_REQUESTED** [review]
This looks wrong, we shouldn't move device mesh like this. I think it should be:
`torch.distributed._tensor.device_mesh -> torch.distributed._device_mesh`
`test/distributed/_tensor/test_device_mesh -> test/distributed/test_device_mesh`

No need to put deprecation warnings in `_tensor.device_mesh`, that file can just import the device mesh from torch.distributed, then we can fix the imports as a separate PR.

File: torch/_dynamo/skipfiles.py [hunk 1]
```diff
@@ -182,7 +182,7 @@ def _module_dir(m: types.ModuleType):
 if torch.distributed.is_available():
     LEGACY_MOD_INLINELIST |= {
         "torch.distributed._tensor.api",
-        "torch.distributed._tensor.device_mesh",
```
Comment (fegin) [inline torch/_dynamo/skipfiles.py:185]: We can keep this line and can avoid confusion when compiling the model that still depend on this path.

**Review by fegin — APPROVED** [review]
LGTM

File: torch/_dynamo/skipfiles.py (hunk 1 above)
Comment (wz337) [reply torch/_dynamo/skipfiles.py:185]: > We can keep this line and can avoid confusion when compiling the model that still depend on this path.

Actually, noob question: I think without this line I am getting the following error:
https://github.com/pytorch/pytorch/blob/7f77ec37be8077079e118d29e209dc202e240d6e/test/dynamo/test_trace_rules.py#L330-L334

Do you know how I can work around this?

**Review by wanchaol — APPROVED** [review]
lgtm, please make sure CI passes :)

**Review by fduwjj — APPROVED** [review]
Actually "we will make torch.distributed._device_mesh public." is not accurate because with `_device_mesh` it is still private? IIUC. But it's great that we decouple it from DTensor.

**Comment by wz337** [comment]
> Actually "we will make torch.distributed._device_mesh public." is not accurate because with `_device_mesh` it is still private? IIUC. But it's great that we decouple it from DTensor.

Ye. We will in following PRs.

**Comment by wz337** [comment]
> Actually "we will make torch.distributed._device_mesh public." is not accurate because with `_device_mesh` it is still private? IIUC. But it's great that we decouple it from DTensor.

I think it should be fine, as the previous path is still importing from the new path so nothing would be broken. 
And since DTensor was prototype as well, we should be able to make any modification we want.

**Comment by fduwjj** [comment]
> I think it should be fine, as the previous path is still importing from the new path so nothing would be broken.

I am not talking about BC breaking here. I mean for new users, they import from the new path right?

**Comment by wz337** [comment]
> > I think it should be fine, as the previous path is still importing from the new path so nothing would be broken.
> 
> I am not talking about BC breaking here. I mean for new users, they import from the new path right?

Once the migration is done, we will add doc to the .rst file and make it public.

**Pull request closed without merging**